KAFKA-9329. KafkaController::replicasAreValid should return error

soondenana opened the pull request:
The `KafkaController::replicasAreValid` method currently returns a
boolean indicating if replicas are valid or not. But the failure
condition loses any context on why replicas are not valid. This change
updates the metod to return the error conition if validation fails. This
allows caller to report the error to the client.

The change also renames the `replicasAreValid` method to
`validateReplicas` to reflect updated semantics.

*More detailed description of your change,
if necessary. The PR title and PR message become
the squashed commit message, so use a separate
comment to ping reviewers.*

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

soondenana committed:
KAFKA-9329. KafkaController::replicasAreValid should return error

The `KafkaController::replicasAreValid` method currently returns a
boolean indicating if replicas are valid or not. But the failure
condition loses any context on why replicas are not valid. This change
updates the metod to return the error conition if validation fails. This
allows caller to report the error to the client.

The change also renames the `replicasAreValid` method to
`validateReplicas` to reflect updated semantics.

soondenana commented:
retest this please

soondenana committed:
Fix scala 2.13 error

The change fixes the error when using map on `scala.collections.Seq` as
2.13 has following change:

> In Scala 2.13 scala.Seq[+A] is an alias for
> scala.collection.immutable.Seq[A], instead of scala.collection.Seq[A]

Fixed code to specify correct collection type as implicit infrence was
creating incorrect collection and failing compilation.

The build still fails with coverage tool failing on deprecation
warnings.

soondenana commented:
retest this please

seanli-rallyhealth commented on core/src/test/scala/integration/kafka/admin/ReassignPartitionsIntegrationTest.scala:
```diff
@@ -37,11 +83,94 @@ class ReassignPartitionsIntegrationTest extends ZooKeeperTestHarness with RackAw
+    servers.map { server =>
+      val port = server.socketServer.boundPort(ListenerName.normalised("PLAINTEXT"))
+      s"localhost:$port"
+    }.headOption.mkString(",")
```
don't think you want .headOption there, which picks up 0 / 1 element from seq of  servers.map { ...}

soondenana replied:
It seems the method to get list of bootstrap server was already there in `TestUtils` class (`TestUtils::bootstrapServers`). Updated code to use that instead.

soondenana committed:
Update test to use TestUtils.bootstrapServers

Couple of tests were duplicating code to create list of bootstrap
servers. There is already a method in TestUtils to do the same, updated
the tests to use that (TestUtils.bootstrapServers)

soondenana commented:
retest this please

soondenana commented:
retest this please

soondenana commented:
retest this please

hachikuji commented:
retest this please

soondenana commented:
2.13 build failed with OOM:

```
20:38:10 [0.036s][warning][os,thread] Failed to start thread - pthread_create failed (EAGAIN) for attributes: stacksize: 1024k, guardsize: 0k, detached.
20:38:10 Error occurred during initialization of VM
20:38:10 java.lang.OutOfMemoryError: unable to create native thread: possibly out of memory or process/resource limits reached
```

2.12 has two unit test failure:

```
kafka.api.GroupEndToEndAuthorizationTest.testNoConsumeWithoutDescribeAclViaSubscribe
kafka.api.PlaintextAdminIntegrationTest.testInvalidAlterPartitionReassignments
```

They pass locally. I will trigger tests again to see if 2.13 can make progress

soondenana commented:
retest this please

hachikuji commented on core/src/main/scala/kafka/controller/KafkaController.scala:
```diff
@@ -1567,13 +1567,15 @@ class KafkaController(val config: KafkaConfig,
-          }
-        } else {
-          reassignmentResults.put(tp, new ApiError(Errors.INVALID_REPLICA_ASSIGNMENT))
+        val errorMessage = targetReplicas.flatMap(validateReplicas(tp, _))
```
nit: `errorMessage` -> `apiError`?

soondenana replied:
Done.

hachikuji commented on core/src/main/scala/kafka/controller/KafkaController.scala:
```diff
@@ -1586,22 +1588,20 @@ class KafkaController(val config: KafkaConfig,
-      case None => true
+  private def validateReplicas(topicPartition: TopicPartition, replicas: Seq[Int]): Option[ApiError] = {
+    val replicaSet = replicas.toSet
+    if (replicas.isEmpty || replicas.size != replicaSet.size)
```
Is it worth checking these cases separately so that we can return a clearer message?

soondenana replied:
Done.

hachikuji commented on core/src/test/scala/integration/kafka/admin/ReassignPartitionsIntegrationTest.scala:
```diff
@@ -37,11 +83,87 @@ class ReassignPartitionsIntegrationTest extends ZooKeeperTestHarness with RackAw
+  }
+
+  @Test
+  def testReassignPartition(): Unit = {
```
Seems like a basic reassignment test case. Do you think it is not covered in `ReassignPartitionsClusterTest`? It's not too clear to me why we have two separate test cases.

soondenana replied:
It is covered. That code uses `zkClient` and `ReassignPartitionsCommand`. This test code uses AdminClient directly. Both of them test same backend code, so maybe its ok to get rid of this as when `ReassignPartitionsCommand` is fixed as part of KIP-500 we will have duplicate test.

Let me know what you prefer and I will take this test out.

soondenana committed:
Take care of review comments

hachikuji commented:
retest this please

soondenana committed:
Fix failing PlaintextAdminIntegrationTest.testInvalidAlterPartitionReassignments

One failure case wasn't handled correctly in
KafkaController::validateReplicas method. Updated code to check the
boolean return condition and generate proper error message.

soondenana commented:
2.13 build had issues when running `kafka.admin.AclCommandTest` and hung. It was killed:

> 22:15:53 Build timed out (after 270 minutes). Marking the build as aborted.

2.12 build had one test failure: `PlaintextAdminIntegrationTest.testInvalidAlterPartitionReassignments`. Pushed a fix for that.

hachikuji commented:
retest this please

hachikuji commented:
ok to test

hachikuji commented on core/src/main/scala/kafka/controller/KafkaController.scala:
```diff
@@ -1588,22 +1590,30 @@ class KafkaController(val config: KafkaConfig,
+      Some(new ApiError(Errors.INVALID_REPLICA_ASSIGNMENT,
+          s"Empty replica list specified in partition reassignment."))
+    else if (replicas.size != replicaSet.size) {
+      val duplicateIds = replicas.groupBy(identity).collect {
```
I think we can remove this and just say there were duplicates in the replica list. Then we can remove the spotbugs warning.

hachikuji commented:
Just a minor comment. Otherwise, looks good.

soondenana committed:
Remove use of collect to avoid spotbugs warning

soondenana commented:
ok to test

soondenana commented:
retest this please

hachikuji commented:
retest this please

hachikuji commented:
ok to test

hachikuji commented:
retest this please

hachikuji reviewed (approved):
LGTM
